Fix checkpoint replay state reset bug

aaudiber opened the pull request:
Without this fix, replaying a checkpoint is almost guaranteed to crash the master. 

The code-level issue is that `mJournalReader.getNextSequenceNumber()` always returns `0` while replaying a checkpoint, so the recently-added
```java
if (mJournalReader.getNextSequenceNumber() == 0) {	
  mMaster.resetState();	
}
```
will reset state before applying every single entry of the checkpoint.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/1863/
Test PASSed.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/1871/
Test FAILed.

aaudiber commented:
@gpang @yuzhu I realized the bug is actually worse than I initially thought - replaying *any* checkpoint will cause the master to crash. I've added a test and updated the PR description

yuzhu commented:
@aaudiber yikes, thanks for catching this..

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/1872/
Test PASSed.

aaudiber commented on core/server/common/src/main/java/alluxio/master/journal/ufs/UfsJournalSnapshot.java:
```diff
@@ -69,23 +69,24 @@ UfsJournalFile getLatestCheckpoint() {
    * @return the journal snapshot
    */
-  static UfsJournalSnapshot getSnapshot(UfsJournal journal) throws IOException {
+  public static UfsJournalSnapshot getSnapshot(UfsJournal journal) throws IOException {
```
the methods are read-only, so I think we need to discourage production code from using them
